src/origin-experiments.js

```diff
@@ -0,0 +1,173 @@
+  }
+
+  /**
+   * Generates an origin experiment token given a config json.
+   * @param {number} version
+   * @param {!JsonObject} json
+   * @param {!webCrypto.CryptoKey} privateKey
+   * @return {!Promise<string>}
+   */
+  generateToken(version, json, privateKey) {
+    const config = stringToBytes(JSON.stringify(json));
+    const data = this.prepend_(version, config);
+    return this.sign_(data, privateKey).then(signature => {
+      return this.append_(data, new Uint8Array(signature));
+    });
+  }
+
+  /**
+   * Verifies an origin experiment token given a public key.
+   * @param {string} token
+   * @param {!Location} location
+   * @param {!webCrypto.CryptoKey} publicKey
+   * @return {!Promise<string>} If token is valid, resolves with the
+   *     experiment ID. Otherwise, rejects with validation error.
+   */
+  verifyToken(token, location, publicKey) {
+    let i = 0;
+    const bytes = stringToBytes(atob(token));
+
+    // Parse version.
+    const version = bytes[i];
+    if (version !== 0) {
+      return Promise.reject(
+          new Error(`Unrecognized token version: ${version}`));
+    }
+    i += 1;
+
+    // Parse config length.
+    const length = bytesToUInt32(bytes.subarray(i, i + LENGTH_BYTES));
+    i += LENGTH_BYTES;
+    if (length > bytes.length - i) {
+      return Promise.reject(new Error(`Unexpected config length: ${length}`));
+    }
+
+    // Parse config itself.
+    const configBytes = bytes.subarray(i, i + length);
+    i += length;
+
+    // Parse unsigned data and its signature.
+    const data = bytes.subarray(0, i);
+    const signature = bytes.subarray(i);
+
+    return this.verify_(signature, data, publicKey).then(verified => {
+      if (!verified) {
+        throw new Error('Failed to verify token signature.');
+      }
+      // Convert config from bytes to JS object.
+      const configStr = bytesToString(configBytes);
+      const config = parseJson(configStr);
+
+      // Check token experiment origin against `location`.
+      const approvedOrigin = parseUrl(config['origin']).origin;
+      const sourceOrigin = getSourceOrigin(location);
+      if (approvedOrigin !== sourceOrigin) {
+        throw new Error(`Config origin (${approvedOrigin}) does not match ` +
+            `window (${sourceOrigin}).`);
+      }
+
+      // Check token expiration date.
+      const experimentId = config['experiment'];
+      const expiration = config['expiration'];
+      if (expiration >= Date.now()) {
+        return experimentId;
+      } else {
+        throw new Error(`Experiment "${experimentId}" has expired.`);
+      }
+    });
+  }
+
+  /**
+   * Returns a byte array: (version + config.length + config)
+   * @param {number} version
+   * @param {!Uint8Array} config
+   * @return {!Uint8Array}
+   */
+  prepend_(version, config) {
+    const data = new Uint8Array(config.length + 5);
+    data[0] = version;
+    new DataView(data.buffer).setUint32(1, config.length, false);
```

Comment:
Hm?

Reply:
Added a comment describing what this does.